test/framework/src/main/java/org/elasticsearch/test/AbstractBuilderTestCase.java

```diff
@@ -212,21 +212,14 @@ public static void afterClass() throws Exception {
 
```

Comment:
Changes on this file are not needed. Can you revert ?

Reply:
Sorry I forgot to add a comment for this. The changes here are because in two of the new test cases I added, the state of `shardContext.mapperService` was changed to force the condition, which affected other test cases as well. By re-creating `serviceHolder` for every test case, side effects from test case does not affect another.

Reply:
Why is it not possible to add your extra fields in `QueryStringQueryBuilderTests#initializeAdditionalMappings` ? That's where we expect tests to modify the shared mapping. Can you show the test failures if that doesn't work ?

Reply:
I did look at that method before. I think the problem of adding additional fields in `QueryStringQueryBuilderTests#initializeAdditionalMappings` was that it is adding additional mappings for ALL test cases in that class (the additional xContent mapping is hard coded in the method, which is an override one as well so I can't change its interface to take in additional parameter easily as well), so that may cause other test cases to fail in that test class as well.

Reply:
Sorry for the delay. I tried to modify `QueryStringQueryBuilderTests#initializeAdditionalMappings` like the following:
````
@Override
    protected void initializeAdditionalMappings(MapperService mapperService) throws IOException {
        XContentBuilder mapping = jsonBuilder().startObject().startObject("_doc").startObject("properties")
            .startObject("prefix_field")
                .field("type", "text")
                    .startObject("index_prefixes").endObject()
                .endObject()
            .startObject("additionalMappedField1")
                .field("type", "text")
            .endObject()
            .startObject("additionalMappedField2")
                .field("type", "text")
            .endObject()
            .startObject("additionalMappedField3")
                .field("type", "text")
            .endObject()
            .startObject("additionalMappedField4")
                .field("type", "text")
            .endObject()
            .endObject().endObject().endObject();

        mapperService.merge("_doc",
            new CompressedXContent(Strings.toString(mapping)), MapperService.MergeReason.MAPPING_UPDATE);
    }
`````
and all tests worked without modification. Could you  try so that we can remove this  unneeded change ?

Reply:
Sorry for the delay. I tried to implement what you suggested and pushed the changes to a different branch / commit, available to review here:  
Branch: https://github.com/zacharymorn/elasticsearch/commits/issue-49002-experiment
Commit: https://github.com/zacharymorn/elasticsearch/commit/e39bcdf27b91dd55c3eb77fbdc7bf04fb1040f8b

I ran the updated test cases in QueryStringQueryBuilderTests and many test cases did fail now because of the uniform initialization of additional fields for each test case:

<img width="1835" alt="Screen Shot 2020-01-08 at 10 10 54 PM" src="https://user-images.githubusercontent.com/2986273/72042942-bd2a3600-3264-11ea-9885-3d2f2c82df69.png">

Am I missing anything here ?

Reply:
Yes sorry I checked without the modified `indices.query.bool.max_clause_count`. It seems tricky to test this behavior here so maybe you can modify or add a new test similar to https://github.com/elastic/elasticsearch/blob/a59b065091a0bec7b9e66ed3ae0e878562947cf8/server/src/test/java/org/elasticsearch/search/query/QueryStringIT.java#L266 ?

Reply:
Didn’t realize there’s an integration test for this as well…I just added a new test there. 

I also removed this line of Index setting https://github.com/elastic/elasticsearch/blob/a59b065091a0bec7b9e66ed3ae0e878562947cf8/server/src/test/java/org/elasticsearch/search/query/QueryStringIT.java#L279-L280 in test `QueryStringIT.testLimitOnExpandedFields` as it doesn’t seems to make sense in the context (why  is it setting INDEX_MAPPING_TOTAL_FIELDS_LIMIT_SETTING  to CLUSTER_MAX_CLAUSE_COUNT + 100?), and have no impact to test result. But please let me know if that’s indeed needed and I can put it back.

Having added the additional test though, I feel this test and the ones I’ve already put in and passing are different levels of tests (integration & unit tests) and thus it doesn’t hurt for them to coexist.  Are you suggesting to have integration test only for these changes?

Reply:
> Are you suggesting to have integration test only for these changes?

Yes, mainly because it seems challenging to have a simple unit test for this and we already have integration tests.

> In test QueryStringIT.testLimitOnExpandedFields as it doesn’t seems to make sense in the context (why is it setting INDEX_MAPPING_TOTAL_FIELDS_LIMIT_SETTING to CLUSTER_MAX_CLAUSE_COUNT + 100?), and have no impact to test result. But please let me know if that’s indeed needed and I can put it back.

I don't think it's needed either but that's unrelated to this pr so not sure if we should remove it now ?

Reply:
I just pushed another commit that reverted the unneeded change to the integration test, and the unit tests and their related changes.

However, I guess I’m still a bit confused for why it’s tricky to keep those unit tests there, when they are already passing and verifying the right logic. As can be seen in the new (reverting) commit, the changes I made to allow unit tests  there are the following:
1. The new unit tests themselves
2. Local, test class level change to force the condition for the added unit tests (setting `indices.query.bool.max_clause_count`)
3. Super test class level change to make sure that each tests are independent to each other so that #2 doesn’t break other test cases (the changes in `AbstractBuilderTestCase.beforeTest`)

These changes combined would allow the newly added test cases to pass, and not break any existing test cases.

I feel I’m still having some gap in understanding the reasoning here, but I’ll leave it to you. Thanks for your patience and time for reviewing my changes!

Reply:
> I feel I’m still having some gap in understanding the reasoning here, but I’ll leave it to you

My reasoning here is that the abstract class for these tests was designed to share the mapping within the tests so it would require a bigger refactoring to change this behavior. We also have a lot of unit tests here so even if they pass I am cautious of the additional work that would be needed if we switch to one configuration per test.